salt/modules/nxos.py

```diff
@@ -57,11 +225,636 @@ def cmd(command, *args, **kwargs):
         salt '*' nxos.cmd show_run
         salt '*' nxos.cmd check_password username=admin password='$5$lkjsdfoi$blahblahblah' encrypted=True
     """
-    proxy_prefix = __opts__["proxy"]["proxytype"]
-    proxy_cmd = ".".join([proxy_prefix, command])
-    if proxy_cmd not in __proxy__:
-        return False
     for k in list(kwargs):
         if k.startswith("__pub_"):
             kwargs.pop(k)
-    return __proxy__[proxy_cmd](*args, **kwargs)
+    local_command = ".".join(["nxos", command])
+    log.info("local command: {}".format(local_command))
+    if local_command not in __salt__:
+        return False
+    return __salt__[local_command](*args, **kwargs)
+
+
+def find(pattern, **kwargs):
+    """
+    Find all instances where the pattern is in the running configuration.
+
+    .. code-block:: bash
+
+        salt '*' nxos.cmd find '^snmp-server.*$'
+
+    .. note::
+        This uses the `re.MULTILINE` regex format for python, and runs the
+        regex against the whole show_run output.
+    """
+    matcher = re.compile(pattern, re.MULTILINE)
+    return matcher.findall(show_run(**kwargs))
+
+
+def get_roles(username, **kwargs):
+    """
+    Get roles assigned to a username.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd get_roles username=admin
+    """
+    user = get_user(username)
+    if not user:
+        return []
+    command = "show user-account {0}".format(username)
+    info = ""
+    info = show(command, **kwargs)
+    if isinstance(info, list):
+        info = info[0]
+    roles = re.search(r"^\s*roles:(.*)$", info, re.MULTILINE)
+    if roles:
+        roles = roles.group(1).strip().split(" ")
+    else:
+        roles = []
+    return roles
+
+
+def get_user(username, **kwargs):
+    """
+    Get username line from switch.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd get_user username=admin
+    """
+    command = 'show run | include "^username {0} password 5 "'.format(username)
+    info = ""
+    info = show(command, **kwargs)
+    if isinstance(info, list):
+        info = info[0]
+    return info
+
+
+def grains(**kwargs):
+    """
+    Get grains for minion.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd grains
+    """
+    if not DEVICE_DETAILS["grains_cache"]:
+        ret = salt.utils.nxos.system_info(show_ver(**kwargs))
+        log.debug(ret)
+        DEVICE_DETAILS["grains_cache"].update(ret["nxos"])
+    return DEVICE_DETAILS["grains_cache"]
+
+
+def grains_refresh(**kwargs):
+    """
+    Refresh the grains for the NX-OS device.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd grains_refresh
+    """
+    DEVICE_DETAILS["grains_cache"] = {}
+    return grains(**kwargs)
+
+
+def sendline(command, method="cli_show_ascii", **kwargs):
+    """
+    Send arbitray commands to the NX-OS device.
+
+    command
+        The command to be sent.
+
+    method:
+        ``cli_show_ascii``: Return raw test or unstructured output.
```

Comment:
is this supposed to be indented 3 maybe?  I think 3 spaces might be normal for sphinx, not sure though.